### ISSUE TITLE
[Inference] Implement Real Sampling Statistics Tracking for Temperature Adjustments and Fallback Strategies

### ISSUE THREAD

**Issue by EffortlessSteven** [EffortlessMetrics/BitNet-rs#373]
## Problem Description

The `GenerationStats` struct in `crates/bitnet-inference/src/generation/autoregressive.rs` contains placeholder fields for critical sampling statistics that are currently hardcoded to `0` instead of being properly tracked. This impacts observability and debugging capabilities for production inference workloads.

## Environment

- **File**: `crates/bitnet-inference/src/generation/autoregressive.rs`
- **Struct**: `GenerationStats` (lines 78-100)
- **Methods**: `get_stats()` (lines 684-739)
- **Build Configuration**: All feature flags (`cpu`, `gpu`)
- **Affected Components**: Autoregressive generation engine, sampling strategies

## Current Implementation (Lines 731-736)

```rust
// Sampling statistics  
temperature_adjustments: 0, // Would track in full implementation
fallback_to_greedy: 0,      // Would track in full implementation  
batched_generations: 0,     // Would track in full implementation

// Quality metrics
average_entropy: 0.0, // Would calculate from logits in full implementation
```

## Root Cause Analysis

The placeholder values indicate that while the infrastructure for sampling statistics exists, the actual tracking logic was never implemented. This creates a gap between the API contract (providing detailed generation statistics) and the actual functionality (returning zeros).

**Impact Assessment:**
- **Severity**: Medium - Affects observability but not core functionality
- **Affected Users**: Production deployments relying on sampling statistics
- **Performance Impact**: Minimal - only adds counters and calculations
- **Business Impact**: Reduced debugging capabilities for inference quality issues

## Proposed Solution

Implement comprehensive tracking for all sampling-related statistics by:

### 1. Add Tracking Fields to AutoregressiveGenerator

```rust
pub struct AutoregressiveGenerator {
    // Existing fields...
    
    // New tracking fields
    temperature_adjustments: usize,
    fallback_to_greedy: usize, 
    batched_generations: usize,
    entropy_samples: Vec<f64>,
}
```

### 2. Initialize Tracking in Constructor

```rust
impl AutoregressiveGenerator {
    pub fn new(/* existing params */) -> Result<Self> {
        Ok(Self {
            // Existing initialization...
            temperature_adjustments: 0,
            fallback_to_greedy: 0,
            batched_generations: 0,
            entropy_samples: Vec::new(),
        })
    }
}
```

### 3. Track Temperature Adjustments

In sampling methods where temperature is dynamically adjusted:

```rust
// When temperature is modified for performance optimization
if self.performance_mode == PerformanceMode::Latency && adjusted_temp != original_temp {
    self.temperature_adjustments += 1;
    log::debug!("Temperature adjusted from {} to {} for latency optimization", 
                original_temp, adjusted_temp);
}
```

### 4. Track Greedy Fallbacks

In sampling fallback logic:

```rust
// When falling back from sampling to greedy decoding
if self.config.do_sample && fallback_to_greedy_triggered {
    self.fallback_to_greedy += 1;
    log::debug!("Falling back to greedy sampling due to performance constraints");
}
```

### 5. Track Batch Operations

In batch processing methods:

```rust
pub async fn process_batch(&mut self, batch_size: usize) -> Result<()> {
    self.batched_generations += 1;
    // Existing batch processing logic...
}
```

### 6. Calculate Average Entropy

```rust
fn calculate_entropy(&self, logits: &[f32]) -> f64 {
    let softmax: Vec<f64> = softmax(logits);
    -softmax.iter()
        .filter(|&&p| p > 0.0)
        .map(|&p| p * p.ln())
        .sum()
}

// In token generation loop
let entropy = self.calculate_entropy(&logits);
self.entropy_samples.push(entropy);
```

### 7. Update get_stats() Implementation

```rust
pub fn get_stats(&self) -> GenerationStats {
    let average_entropy = if !self.entropy_samples.is_empty() {
        self.entropy_samples.iter().sum::<f64>() / self.entropy_samples.len() as f64
    } else {
        0.0
    };

    GenerationStats {
        // Existing fields...
        temperature_adjustments: self.temperature_adjustments,
        fallback_to_greedy: self.fallback_to_greedy,
        batched_generations: self.batched_generations,
        average_entropy,
        // Remaining fields...
    }
}
```

### Phase 1: Core Infrastructure
- [ ] Add tracking fields to `AutoregressiveGenerator` struct
- [ ] Update constructor to initialize tracking fields
- [ ] Implement entropy calculation helper function
- [ ] Update `get_stats()` to return real values

### Phase 2: Integration Points
- [ ] Identify all temperature adjustment locations in sampling code
- [ ] Identify all greedy fallback locations in sampling strategy
- [ ] Identify all batch processing locations
- [ ] Add tracking increments at each location

### Phase 3: Quality Assurance
- [ ] Add unit tests for tracking accuracy
- [ ] Add integration tests with different sampling configurations
- [ ] Verify tracking works across CPU and GPU backends
- [ ] Performance benchmark to ensure minimal overhead

### Phase 4: Documentation and Observability
- [ ] Update API documentation for GenerationStats fields
- [ ] Add logging for significant sampling events
- [ ] Create example demonstrating statistics usage
- [ ] Update health endpoint to expose sampling metrics

### Unit Tests
```rust
#[test]
fn test_temperature_adjustment_tracking() {
    // Test that temperature adjustments are properly counted
}

#[test] 
fn test_greedy_fallback_tracking() {
    // Test that greedy fallbacks are properly counted
}

#[test]
fn test_entropy_calculation() {
    // Test entropy calculation accuracy with known distributions
}
```

### Integration Tests
```rust
#[test]
fn test_sampling_stats_cpu_backend() {
    // Test end-to-end statistics tracking on CPU
}

#[test]  
fn test_sampling_stats_gpu_backend() {
    // Test end-to-end statistics tracking on GPU
}
```

### Performance Tests
- Measure overhead of statistics tracking (should be <1% impact)
- Verify memory usage remains bounded for long generations
- Test with various sampling configurations (greedy, nucleus, top-k)

## Risk Assessment

**Potential Complications:**
- **Thread Safety**: Ensure tracking is safe in concurrent contexts
- **Memory Growth**: Prevent unbounded growth of entropy samples
- **Performance Impact**: Minimize computational overhead
- **Backward Compatibility**: Ensure existing API contracts are maintained

**Mitigation Strategies:**
- Use atomic counters for thread-safe tracking
- Implement sliding window for entropy samples (e.g., last 1000 tokens)
- Make tracking optional via feature flag if performance is critical
- Maintain existing GenerationStats field names and types

### Functional Requirements
- [ ] All GenerationStats fields return real values instead of placeholders
- [ ] Temperature adjustments are accurately tracked across all sampling modes
- [ ] Greedy fallbacks are counted when sampling falls back to deterministic selection
- [ ] Batch operations are properly tracked
- [ ] Average entropy is calculated correctly from logits distributions

### Quality Requirements  
- [ ] Performance overhead is less than 1% for typical workloads
- [ ] Memory usage remains bounded (O(1) or O(k) for fixed window)
- [ ] Thread safety is maintained in concurrent generation scenarios
- [ ] All tracking is covered by comprehensive unit tests

### Documentation Requirements
- [ ] GenerationStats fields are properly documented with examples
- [ ] API documentation explains when each counter is incremented
- [ ] Example code demonstrates how to interpret sampling statistics

## Related Issues

This issue addresses the sampling statistics tracking gap identified in production readiness analysis. It complements:

- Model validation and accuracy tracking improvements
- Production inference server observability enhancements  
- Performance monitoring and optimization efforts

The implementation should coordinate with any concurrent work on:
- Sampling strategy improvements
- Performance optimization in generation loops
- Health endpoint and monitoring infrastructure

## References

- BitNet.rs sampling implementation: `crates/bitnet-inference/src/generation/sampling.rs`
- Autoregressive generation: `crates/bitnet-inference/src/generation/autoregressive.rs`
- Performance monitoring: `docs/performance-benchmarking.md`
- Health endpoints: `docs/health-endpoints.md`
